packages/lodestar/src/sync/backfill/backfill.ts

```diff
@@ -0,0 +1,220 @@
+  private processor = new ItTrigger();
+  private peers = new PeerSet();
+
+  constructor(modules: BackfillSyncModules, opts?: BackfillSyncOpts) {
+    super();
+    this.chain = modules.chain;
+    this.network = modules.network;
+    this.db = modules.db;
+    this.config = modules.config;
+    this.logger = modules.logger;
+    this.opts = opts ?? {batchSize: BATCH_SIZE};
+    this.network.events.on(NetworkEvent.peerConnected, this.addPeer);
+    this.network.events.on(NetworkEvent.peerDisconnected, this.removePeer);
+
+    this.sync()
+      .then((oldestSlotSynced) => {
+        this.emit(BackfillSyncEvent.completed, oldestSlotSynced);
+        this.logger.info("BackfillSync completed", {oldestSlotSynced});
+        // Sync completed, unsubscribe listeners and don't run the processor again
+        // Backfill is never necessary again until the node shuts down
+        this.close();
+      })
+      .catch((e) => {
+        if (e instanceof ErrorAborted) {
+          return; // Ignore
+        }
+        this.logger.error("BackfillSync processor error", e);
+      });
+  }
+
+  /** Throw / return all AsyncGenerators */
+  close(): void {
+    this.network.events.off(NetworkEvent.peerConnected, this.addPeer);
+    this.network.events.off(NetworkEvent.peerDisconnected, this.removePeer);
+    this.processor.end(new ErrorAborted("BackfillSync"));
+  }
+
+  /** Returns oldestSlotSynced */
+  private async sync(): Promise<Slot> {
+    const {root, epoch} = this.chain.getFinalizedCheckpoint();
+    if (epoch <= 0) {
+      this.logger.debug("BackfillSync no necessary on genesis");
+      return 0;
+    }
+
+    this.logger.debug("BackfillSync initializing from cp", {epoch, root: toHexString(root)});
+
+    // Try to figure out if BackfillSync is necessary. Note that this check is very inexact
+    // - Define "target range" to backfill as: [oldestSlotRequired, cpStartSlot]
+    // - If there's at least 1 block in target range, consider yourself BackfillSynced
+    const cpStartSlot = computeStartSlotAtEpoch(epoch);
+    const oldestSlotRequired = this.getOldestRequiredSlot();
+    const slots = await this.db.blockArchive.keys({gte: oldestSlotRequired, lte: cpStartSlot, limit: 1});
+    if (slots.length > 0) {
+      this.logger.debug("BackfillSync has at least one block in range, skipping", {oldestSlotRequired, cpStartSlot});
+      return slots[0];
+    }
+
+    // Attempt to find the anchor block in the DB
+    try {
+      this.anchorBlock = await this.db.blockArchive.getByRoot(root);
+    } catch (e) {
+      this.logger.error("Error fetching blockArchive", {root: toHexString(root)}, e);
+    }
+
+    // Start processor loop
+    this.processor.trigger();
+
+    for await (const _ of this.processor) {
+      const peer = shuffleOne(this.peers.values());
+      if (!peer) {
+        this.logger.debug("BackfillSync no peers yet");
+        continue;
+      }
+
+      try {
+        // First get anchor block
```

Comment:
Re-organized to first try to get anchorBlock, then sync the range of blocks. Allows to organize the completedness check is a clearer way